Add tooling to enable quick release testing

amcp opened the pull request:
This change adds a command line tool that ideally will allow any combination of tests to be run. All you have to do is define the command line arguments to maven for each test run and execute the CLI as per the README.md. Currently, the maximum number of parallel actions (build steps) in a stage is 5, so I split up the various test targets into groups of up to four builds.

Status of builds in j1:
- [x] bdb - Passed in 15 min 10 secs
- [x] hadoop-lucene-solr - Passed in 21 minutes 12 seconds - artifact assembly needs to be fixed 
- [x] test - Passed in 21 minutes 37 seconds
- [x] cassandra - Passed in 110 minutes

Status of builds in j2:
- [x] hbase098 - Passed in 65 minutes
- [x] hbase10 - Passed in 39 minutes 44 seconds
- [ ] es - Failed in 15 min 12 sec - cant be run as root
- [x] cql - Failed in 67 minutes (cql tests will be fixed elsewhere)

Status of builds in jtp1:
- [x] es-hadoop-lucene-solr - Passed in 6 minutes 49 seconds - artifact assembly needs to be fixed 
- [x] bdb - Passed in 118 minutes

Status of builds in jtp2:
- [x] test - Passed in 32 minutes 4 seconds
- [x] cassandra - Failed in 131 minutes (known failures)
- [x] hbase098 - Failed in 409 minutes (known failures)
- [x] hbase10 - Passed in 467 minutes (very close to timeout limit of 8 hours)
- [x] cql - Timed out in 8 hours (currently 14-15 hours. the timeout is a limitation of CodeBuild. will not run to completion until docker orchestration drops the run time for cql)

amcp commented:
@sjudeng @jerryjch I am running off amcp/janusgraph master now. I will report back how long it took for all pipelines to finish testing. But, the PR is good to start reviewing.

amcp commented:
I noticed that the builds will time out so I kicked off a new build with the maximum limit supported, 8 hours. 
<img width="962" alt="pipeline" src="https://cloud.githubusercontent.com/assets/1066771/23179755/5206e0f6-f8b2-11e6-9374-19203d589708.png">

amcp commented:
TP pipeline number 1, below
<img width="948" alt="screen shot 2017-02-22 at 3 53 53" src="https://cloud.githubusercontent.com/assets/1066771/23179836/a36be6da-f8b2-11e6-8df2-858c597ca706.png">

amcp commented:
@amcp regroup tests into those that pass and those that fail. Want to see the kind of integration that can be had on passing pipelines.

mbrukman commented:
@amcp — you used two different addresses here, one as "author", a different one as "committer", so your CLA is not recognized. You can see this because there are two photos next to each of your commits. Could you please fix this?

amcp commented:
@mbrukman I will fix my commit.

sjudeng commented on .gitignore:
```diff
@@ -13,6 +13,7 @@ target/
 /log/
 /output/
 /scripts/
+/janusgraph-codepipelines-ci/dependency-reduced-pom.xml
```
Can you update Maven shade plugin config in janusgraph-codepipelines-ci/pom.xml to include createDependencyReducedPom=false and then remove this exclusion?

amcp replied:
fixed

sjudeng commented on .gitignore:
```diff
@@ -26,6 +27,9 @@ target/
+/codepipelines-ci/.classpath
+/codepipelines-ci/.project
+/codepipelines-ci/.settings/
 /janusgraph-*/.classpath
```
These exclusions aren't necessary since they're covered under the below janusgraph-* patterns

amcp replied:
fixed

sjudeng commented on buildspec.yml:
```diff
@@ -0,0 +1,17 @@
+artifacts:
+  base-directory: module
+  files:
+    - target/**/*
```
Can you move this to the janusgraph-codepipelines-ci module?

amcp replied:
The standard location is at the [root of the repository](http://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-name-storage). We could add a step in the manual workflow to copy the file to the repo root, zip up the repo, copy it to S3, and use that as the source for Code Pipelines. Sounds like a lot of mechanism. Can we revisit this later, perhaps adding a TODO?

sjudeng commented on janusgraph-berkeleyje/src/test/resources/release-test-groups/bdb1:
```diff
@@ -0,0 +1 @@
+org.apache.tinkerpop.gremlin.process.TraversalPerformanceTest
```
Where is this test resource used? Same question for bdb2 and bdb3 below.

amcp replied:
Now that I have [CI working for the DynamoDB Storage Backend here](https://travis-ci.org/awslabs/dynamodb-titan-storage-backend/builds/237101582?utm_source=github_status&utm_medium=notification), I want to build on this experience to inform the mechanism for dividing up tests here.

Basically, considering the large number of tests we have, parameterization and the length of concatenated canonical test class names, using these kinds of files is unwieldy and potentially may cause length exceeded errors on the command line used to start tests. I would rather define test categories and use those to drive the tests.

amcp replied:
Of course, depending on the granularity of division, we may need a tool to make sure that TP updates and janusgraph-test updates did not introduce new tests that are not labelled with the new test categories.

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,28 @@
+--github-repo janusgraph \
+--github-branch <branch you want to test> \
+--profile code-pipelines \
+--environments=./environments.json \
```
Should --environment be --pipelines and can you document the expected JSON content format?

amcp replied:
renamed and moved to the bottom

sjudeng commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,28 @@
+```bash
+mvn clean && mvn package && java -jar target/janusgraph-codepipelines-ci-0.1.0-SNAPSHOT.jar --region <region close to you> \
+--bucket <bucket> \
+--name <name of ci stack> \
````
--name is not a supported CLI argument

sjudeng replied:
Are github-owner and github-token really required? I'm not familiar with AWS CodePipelines but the below docs suggest personal access tokens aren't required when using Git pull. If those docs are applicable it looks like only OutputBucketName and AllowedIps (GitHub 192.30.252.0/22?) would be required. I'm not positive but it doesn't look like even the ApiSecret would be required.

https://aws.amazon.com/blogs/devops/integrating-git-with-aws-codepipeline/

amcp replied:
fixed

amcp replied:
Thank you for referring me to the blog article. My read on the article indicates that OutputBucketName plus one of AllowedIps or ApiSecret would be required. But, I like your inclination. AllowedIps requires less effort on the user's part and should continue to be viable for the foreseeable future.

amcp replied:
On second thought, the existing configuration sets up a hook where new commits on the selected branch automatically get built in code pipeline/codebuild. Scenario: set up the pipeline for a release branch, tests get run once and some fail, you fix the failing tests, push a new commit to the release branch and then codepipeline/codebuild automatically builds anew.

sjudeng commented on janusgraph-codepipelines-ci/pipe.json:
```diff
@@ -0,0 +1,15 @@
+        "cassandra" : { "MODULE" : "janusgraph-cassandra" },
+        "es" : { "MODULE" : "janusgraph-es", "ARGS" : "-DthreadCount=1" },
+        "hadoop" : { "MODULE" : "janusgraph-hadoop-parent/janusgraph-hadoop-2" },
+        "hbase09" : { "MODULE" : "janusgraph-hbase-parent/janusgraph-hbase-098" }
```
/"hbase09"/"hbase098"/?

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/pipe.json:
```diff
@@ -0,0 +1,15 @@
+{
+    "j1pass" : {
+        "bdb" : { "MODULE" : "janusgraph-berkeleyje" },
+        "lucene" : { "MODULE" : "janusgraph-lucene", "ARGS" : "-DthreadCount=1" },
```
threadCount=1 originated during some Travis debugging but I don't think it ever actually solved any problems. I'd suggest removing it here and throughout.

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+    <url>http://janusgraph.org</url>
+    <properties>
+        <top.level.basedir>${basedir}/..</top.level.basedir>
+        <aws.sdk.version>1.11.86</aws.sdk.version>
```
Move to parent pom?

amcp replied:
I want to limit exposure to AWS specific APIs to the specific places where it is used / can be used (AWS ES, HBase on EMR, and now, CodePipelines). Currently the only code dependency is in janusgraph-codepipelines-ci package, so lets keep the SDK version here for now. If we find we need the SDK in more than one place, then I think that is the time to move it up into the parent POM.

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+    <properties>
+        <top.level.basedir>${basedir}/..</top.level.basedir>
+        <aws.sdk.version>1.11.86</aws.sdk.version>
+        <guava.version>21.0</guava.version>
```
Can you use the projec guava version (18.0)?

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+        <aws.sdk.version>1.11.86</aws.sdk.version>
+        <guava.version>21.0</guava.version>
+        <project.http.version>1.22.0</project.http.version>
+        <maven-compiler-plugin.version>3.6.1</maven-compiler-plugin.version>
```
Are these properties needed?

amcp replied:
not used so i removed them.

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+            <artifactId>aws-java-sdk-sts</artifactId>
+            <version>${aws.sdk.version}</version>
+        </dependency>
+        <dependency>
```
Doesn't appear that aws-java-sdk-sts is used

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+            <artifactId>lombok</artifactId>
+            <version>1.16.14</version>
+            <scope>provided</scope>
+        </dependency>
```
lombok appears unused

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
janusgraph-test is unused

amcp replied:
removed

amcp replied:
actually, it is needed because `mvn install` in the codepipelines-ci directory fails saying the PerformanceTests.class category is missing. I guess I could override the test configuration inherited from the parent pom but that sounds like a lot of work so I prefer just to depend on janusgraph-test.

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+                <directory>${basedir}/src/test/resources</directory>
+            </testResource>
+        </testResources>
+
```
There aren't any src/main/resources. Are these definitions needed?

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
Is test-jar needed?

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
```
Correct indentation

sjudeng replied:
If possible add configuration for createDependencyReducedPom=false and remove top level .gitignore of same

amcp replied:
fixed the indentation and added createDependencyReducedPom=false

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+                </plugins>
+            </build>
+        </profile>
+    </profiles>
```
Shouldn't need this as janusgraph-codepipelines-ci jar shouldn't be included in release.

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+                <version>2.8.1</version>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
```
Unnecessary dependencyManagement definition

amcp replied:
Through TinkerPop, I think the wrong versions of SDK dependencies were being pulled in. Hence the desire to limit scope of SDK dependency and to manage the versions of SDK dependencies here.

sjudeng commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,290 @@
+    public static final String ACTION = "Action";
+    public static final String VERSION = "Version";
+    public static final String VERSION_DATE = "2012-10-17";
+    public static final String STATEMENT = "Statement";
```
The above constants don't appear to be used. Can they be removed?

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,290 @@
+    public static final String VERSION = "Version";
+    public static final String VERSION_DATE = "2012-10-17";
+    public static final String STATEMENT = "Statement";
+    public static final String AWS_CODEBUILD_JAVA_OPENJDK_8 = "aws/codebuild/java:openjdk-8";
```
Should this be a property?

amcp replied:
JDK 9 is coming soon so I guess we can pull it out as a parameter.

amcp replied:
And maybe we want to run with a different JDK or environment. Anyway, ill pull it out.

sjudeng commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,290 @@
+import java.util.stream.Collectors;
+
+/**
+ * Created by amcp on 2017/02/20.
```
Replace with `@author` for consistency with documentation elsewhere in the project

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,290 @@
+                                                               String artifactsName,
+                                                               ComputeType computeType,
+                                                               String computeImage) {
+        System.out.println("Creating CodeBuild project for " + envName);
```
Remove System.out.printlns throughout. Use slf4j logger.

amcp replied:
removed here and elsewhere

sjudeng commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,290 @@
+            createRequiredOneArgOption("pipelines" /*longOpt*/,"Path to JSON file containing abbreviated pipeline definitions");
+    private static final Option CODE_BUILD_SERVICE_ROLE_ARN_OPTION =
+            createRequiredOneArgOption("codebuild-role-arn" /*longOpt*/,
+                    "ARN of the service role for CodeBuild (http://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html#setting-up-service-role)");
```
Move these option definitions to the top with other field definitions (unless there's precedent elsewhere for breaking up definitions like this)

sjudeng replied:
Also recommend removing `/*longopt*/`, `/*hasArg*/`, `/*shortArg*/`, etc. inline comments. Some IDEs (e.g. IntelliJ) already show argument names making this redundant.

amcp replied:
moved to the top and removed the inline comments.

sjudeng commented on janusgraph-codepipelines-ci/src/test/java/org/janusgraph/codepipelines/TestContainerOne.java:
```diff
@@ -0,0 +1,29 @@
+import org.junit.Test;
+
+/**
+ * Created by amcp on 2017/02/24.
```
Replace with `@author`

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/src/test/java/org/janusgraph/codepipelines/TestContainerOne.java:
```diff
@@ -0,0 +1,29 @@
+            //noop
+        }
+    }
+}
```
What are these tests testing? Question applies to TestContainerTwo below as well.

amcp replied:
They are sample classes for me to run and make sure that I was able to select using command line arguments in the JSON files.

sjudeng commented on janusgraph-codepipelines-ci/src/test/resources/excludes:
```diff
@@ -0,0 +1 @@
+**/*.java*
```
Where is this test resource used? Likewise for longTests1 and longTests2 below.

sjudeng commented on janusgraph-codepipelines-ci/tp-pipe.json:
```diff
@@ -0,0 +1,15 @@
+    "jtp1pass" : {
+    	"es-tp" : { "MODULE" : "janusgraph-es", "ARGS" : "-DthreadCount=1 -Dtest.skip.tp=false -DskipTests=true" },
+    	"hadoop-tp" : { "MODULE" : "janusgraph-hadoop-parent/janusgraph-hadoop-2", "ARGS" : "-Dtest.skip.tp=false -DskipTests=true" },
+        "lucene-tp" : { "MODULE" : "janusgraph-lucene", "ARGS" : "-DthreadCount=1 -Dtest.skip.tp=false -DskipTests=true" },
```
See above regarding removing threadCount=1.

amcp replied:
removed threadCount

sjudeng commented on janusgraph-codepipelines-ci/tp-pipe.json:
```diff
@@ -0,0 +1,15 @@
+    	"es-tp" : { "MODULE" : "janusgraph-es", "ARGS" : "-DthreadCount=1 -Dtest.skip.tp=false -DskipTests=true" },
+    	"hadoop-tp" : { "MODULE" : "janusgraph-hadoop-parent/janusgraph-hadoop-2", "ARGS" : "-Dtest.skip.tp=false -DskipTests=true" },
+        "lucene-tp" : { "MODULE" : "janusgraph-lucene", "ARGS" : "-DthreadCount=1 -Dtest.skip.tp=false -DskipTests=true" },
+        "solr-tp" : { "MODULE" : "janusgraph-solr", "ARGS" : "-DthreadCount=1 -Dtest.skip.tp=false -DskipTests=true" }
```
Remove extra leading space on solr-tp and lucene-tp lines

amcp replied:
alignment is fixed

sjudeng commented on janusgraph-codepipelines-ci/tp-pipe.json:
```diff
@@ -0,0 +1,15 @@
+    "jtp2fail" : {
+        "bdb-tp" : { "MODULE" : "janusgraph-berkeleyje", "ARGS" : "-Dtest.skip.tp=false -DskipTests=true" },
+        "cassandra-tp" : { "MODULE" : "janusgraph-cassandra", "ARGS" : "-Dtest.skip.tp=false -DskipTests=true" },
+        "hbase09-tp" : { "MODULE" : "janusgraph-hbase-parent/janusgraph-hbase-098", "ARGS" : "-Dtest.skip.tp=false -DskipTests=true" },
```
/"hbase09-tp"/"hbase-098-tp"/?

amcp replied:
fixed

sjudeng commented on pom.xml:
```diff
@@ -114,10 +114,14 @@
         <compiler.target>1.8</compiler.target>
+        <test.pipelines.excludes></test.pipelines.excludes>
+        <test.pipelines.includes></test.pipelines.includes>
+        <test.skip.pipelines>true</test.skip.pipelines>
```
Where are these used?

amcp replied:
I guess nowhere so i removed them.

sjudeng commented on pom.xml:
```diff
@@ -276,6 +280,7 @@
                             <exclude>**/*ProcessPerformanceTest.java</exclude>
                             <exclude>**/*StructureTest.java</exclude>
                         </excludes>
+                        
```
Remove extra newline

amcp replied:
fixed

sjudeng commented on pom.xml:
```diff
@@ -307,16 +312,39 @@
+                                    <log4j.configuration>file:${project.build.directory}/test-classes/log4j.properties</log4j.configuration>
+                                </systemPropertyVariables>
+                            </configuration>
+                        </execution>
```
Do the janusgraph-codepipeline-ci tests warrant this custom test configuration? Would default surefire configuration suffice?

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/pom.xml:
```diff
@@ -0,0 +1,171 @@
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>janusgraph-codepipelines-ci</artifactId>
+    <name>JanusGraph-CodePipeline CI: Distributed Graph Database</name>
```
Update description from "Distributed Graph Database" to something describing module (e.g. "AWS Codepipelines" or something)

amcp replied:
updated

amcp commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,311 @@
+ */
+@Slf4j
+public class AwsCodePipelinesCi {
+    private static final String AWS_CODEBUILD_JAVA_OPENJDK_8 = "aws/codebuild/java:openjdk-8";
```
pull into a parameter

amcp commented on janusgraph-codepipelines-ci/src/main/java/org/janusgraph/codepipelines/AwsCodePipelinesCi.java:
```diff
@@ -0,0 +1,311 @@
+                        .withPackaging(ArtifactPackaging.NONE)
+                        .withType(ArtifactsType.CODEPIPELINE)
+                        .withName(artifactsName))
+                .withTimeoutInMinutes(480) //8 hours max - TODO externalize
```
pull into a parameter

amcp commented:
@sjudeng I'm working on this again. Any help you could give to push this to completion would be greatly appreciated.

amcp commented:

14:29:59
SLF4J: Class path contains multiple SLF4J bindings.

14:29:59
SLF4J: Found binding in [jar:file:/root/.m2/repository/org/slf4j/slf4j-log4j12/1.7.12/slf4j-log4j12-1.7.12.jar!/org/slf4j/impl/StaticLoggerBinder.class]

14:29:59
SLF4J: Found binding in [jar:file:/root/.m2/repository/ch/qos/logback/logback-classic/1.1.2/logback-classic-1.1.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]

14:29:59
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.

14:29:59
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]

14:29:59
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: -1,924,987.913 sec <<< FAILURE! - in org.janusgraph.diskstorage.hbase.HBaseDistributedStoreManagerTest

14:29:59
org.janusgraph.diskstorage.hbase.HBaseDistributedStoreManagerTest Time elapsed: -1,924,987.914 sec <<< ERROR!

14:29:59
java.lang.RuntimeException: java.io.IOException: Cannot run program "../janusgraph-hbase-098/bin//hbase-daemon.sh": error=13, Permission denied

14:29:59
at java.lang.UNIXProcess.forkAndExec(Native Method)

14:29:59
at java.lang.UNIXProcess.<init>(UNIXProcess.java:187)

14:29:59
at java.lang.ProcessImpl.start(ProcessImpl.java:134)

14:29:59
at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)

14:29:59
at org.janusgraph.HBaseStorageSetup.runCommand(HBaseStorageSetup.java:263)
at org.janusgraph.HBaseStorageSetup.runCommand(HBaseStorageSetup.java:263)

14:29:59
at org.janusgraph.HBaseStorageSetup.startHBase(HBaseStorageSetup.java:135)

14:29:59
at org.janusgraph.diskstorage.hbase.HBaseDistributedStoreManagerTest.startHBase(HBaseDistributedStoreManagerTest.java:34)

amcp commented:
@sjudeng The HBase error is fixed. I updated test status above. I am stuck on getting ES tests to run. I think we removed embedded mode, and now ES tests are looking for an ES endpoint and not finding it. Can you point me to the right way to start ES tests?

amcp commented:
For TP Cassandra tests, 4/10 tests in TraversalInterruptionComputerTest are failing:
1. shouldRespectThreadInterruptionInVertexStep[expectInterruption(g_V_out)]
2. shouldRespectThreadInterruptionInVertexStep[expectInterruption(g_V_in)]
3. shouldRespectThreadInterruptionInVertexStep[expectInterruption(g_V_inE)]
4. shouldRespectThreadInterruptionInVertexStep[expectInterruption(g_E_inV)]

sjudeng commented:
I've been working on another PR (your #148) and haven't gotten back to this yet. I'll try to do so in the coming days. But regarding your questions above ...

The interrupt tests have been failing sporadically I think since the update to TP 3.2.4. I created #258 to track resolution.

Local test ES instance is currently managed through the Elasticsearch distribution artifact as follows.

1. Include the [org.elasticsearch.distribution.zip:elasticsearch](https://github.com/JanusGraph/janusgraph/blob/9ad315a2ab00b091634b85bc9b268b05b1a9422f/janusgraph-es/pom.xml#L58-L74) dependency
2. [Unpack](https://github.com/JanusGraph/janusgraph/blob/9ad315a2ab00b091634b85bc9b268b05b1a9422f/janusgraph-es/pom.xml#L166-L183) distribution zip into the project build directory
3. Manage instance (e.g. [start](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticsearchRunner.java#L112)) based on unpacked [distribution path](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticsearchRunner.java#L64).

Alternatively you can use an externally managed ES instance for testing in which case you'd include the `is.external.es` property when running tests.

```bash
mvn clean install -pl janusgraph-es -Dis.external.es=true
```

amcp commented:
@sjudeng Regarding ES, are you saying that just running verify should be enough to take advantage of local test ES instance machinery? I wonder if I need to open up some permissions as ElasticsearchRunner calls `runCommand`

sjudeng commented:
Yes all tests run automatically through Maven. Are you running the tests as a non-root user? Elasticsearch won't start under root. Otherwise if you have access to the filesystem you can check the log, `janusgraph-es/target/elasticsearch-5.2.2/logs/elasticsearch.log`.

amcp commented:
I am currently running all tests as a root user. The PrivilegedMode configuration is set at the build project level so I will have to pull it into the JSON files.

amcp commented:
@sjudeng I wonder if there is a way to get the cql TP tests to run as fast as the cassandra TP tests?

sjudeng commented:
I thought the CQL TP tests run with similar timing as Cassandra/HBase.

amcp commented:
Can you point me to the results of your last run of CQL TP tests? Maybe I am missing some command line arguments or properties.

sjudeng commented:
See [this test summary](https://github.com/JanusGraph/janusgraph/files/1070207/test_tp32_cql_summary_20170505.txt) but note it's pre #244 so runtime was closer to 20 hours. Call was something like `mvn clean install -Dtest.skip.tp=false -fn -pl janusgraph-cql`. Do the default non-TP tests run without timing out?

amcp commented:
@sjudeng All the default non-TP tests run without timing out now. See the PR description for latest timings, but the longest one is cassandra at 110 minutes.

sjudeng commented:
I ran the full suite again and janusgraph-cql did take 15 hours to finish (compared to 4 hours for janusgraph-cassandra) so that confirms what you're seeing.

```
[INFO] JanusGraph-Test: Test Suite for JanusGraph ........ SUCCESS [41:17.523s]
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database . SUCCESS [2:18:10.966s]
[INFO] JanusGraph-Cassandra: Distributed Graph Database .. SUCCESS [4:18:21.171s]
[INFO] JanusGraph-CQL: Distributed Graph Database ........ SUCCESS [15:03:15.787s]
[INFO] JanusGraph-HBase: 1.x Compatibility Shim .......... SUCCESS [8:55:16.353s]
[INFO] JanusGraph-Hadoop: 2.x Compatibility Shim ......... SUCCESS [7:02.152s]
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support  SUCCESS [11:02.612s]
[INFO] JanusGraph-Lucene: Indexing Support ............... SUCCESS [1:56.719s]
[INFO] JanusGraph-Solr: Distributed Indexing Support ..... SUCCESS [8:46.541s]
```

amcp commented:
@sjudeng So Cassandra TP takes 4h18 and CQL takes 15h3 time to run. CQL TP tests take 3.5 times more time than Cassandra. So, if the Cassandra heavy hitters are TraversalInterruptionComputerTest, MatchTest$GreedyMatchTraversals and MatchTest$CountMatchTraversals and GroupTest$Traversals, maybe we can split up CQL so that it runs within the 8 hour time limit.

amcp commented:
@sjudeng Also, this PR is getting super long, can we close it and create a new one?

sjudeng commented:
I think it'll be tricky to split up the tests because of TinkerPop testing framework abstractions. It might be worth trying to understand why the CQL tests are running so much longer.

@pkendall64 or @samant-orionhealth do you agree something doesn't look right with these numbers and any ideas? First thing I looked at was your ID batch size fix but that update has already been applied.

pkendall64 commented:
@sjudeng The only reason I can see that it would be that much slower is the connection to cassandra being torn down and re-created for each test. I'll re-run them here and see if I can track it down.

sjudeng commented:
@pkendall64 Thanks for taking a look. It would be good even just to confirm that the tests suites finish much faster when Cassandra is managed externally and not restarted for each test. I think we'd have some decent options in this case (e.g. manage Cassandra Docker container as part of testing).

amcp commented:
@sjudeng seems like the cql TP test timeout will be addressed elsewhere as it requires docker orchestration. Only thing left here is to get es test passing and to fix the way I am assembling the composite builds.

amcp commented:
@sjudeng getting the cql tests passing is also out of scope of this cr, so I am marking those complete as well.

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+mvn clean package
+#define variables
+export AWS_PROFILE_NAME='profile_to_use'
+export AWS_ACCOUNT_NUMBER='000000000000'
```
Is this the root account number?

amcp replied:
I can use STS to get the root account number, I will update

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+--codebuild-role-arn arn:aws:iam::${AWS_ACCOUNT_NUMBER}:role/${AWS_CODEBUILD_ROLE} \
+--codepipeline-role-arn arn:aws:iam::${AWS_ACCOUNT_NUMBER}:role/${AWS_CODEPIPELINE_ROLE} \
+--github-token ${GITHUB_TOKEN} \
+--pipelines ${PWD}/pipe.json
```
Consider moving pipelines file to the environment as well.

amcp replied:
I provided two pipelines files so I prefer that users be able to copy paste the execution command after setting up their environment.

amcp replied:
second thought, i like your comment below so I will change according to your suggestion

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+--github-token ${GITHUB_TOKEN} \
+--pipelines ${PWD}/pipe.json
+
+#kick off tinkerpop tests
```
Consider removing the second call here. It's a complicated call and the only difference is the pipeline used. Instead create another section or discussion (or link) on how to create the pipelines files and indicate that examples are included to run default (pipe.json) and TinkerPop (tp-pipe.json) tests.

amcp replied:
done

sjudeng commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,57 @@
+```
+8. Navigate to the [AWS Console](https://console.aws.amazon.com) and check on the status of your pipeline to see status.
+
+## Future ideas
````
What about tracking this under a separate issue instead of here?

amcp replied:
moved to an issue #367

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+#JanusGraph CodePipelines CI
```
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+CodePipelines CI is a mechanism JanusGraph can use to do release testing in massively
+parallel fashion (to the extent of AWS CodePipelines and CodeBuild service limits).
+
+## Steps to kick off a build
```
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+It also requires you to create two service roles in IAM: one for CodePipeline and
+one for CodeBuild.
+
+1. Get a personal access token from [GitHub](https://github.com/settings/tokens) with repo and admin:repo_hook scopes.
```
I've confirmed in a [use case](https://github.com/stelligent/dromedary/blame/master/README.md#L77) and The [documentation](http://docs.aws.amazon.com/codepipeline/latest/userguide/integrations-action-type.html) only mentions the repo scope but it assumes the user will create the pipeline using the console. To create the pipeline programmatically, the admin:repo_hook scope is required.

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+one for CodeBuild.
+
+1. Get a personal access token from [GitHub](https://github.com/settings/tokens) with repo and admin:repo_hook scopes.
+2. Navigate to the [AWS Console](https://console.aws.amazon.com) and create an IAM User with the following managed policies: AmazonS3FullAccess, AWSCodePipelineFullAccess, AWSCodeBuildAdminAccess.
```
It is possible to do with the AWS CLI, but to maintain the chain of trust and accountability, it is preferable that the user manually creates the user. The policies are managed, so it is just point and click.

amcp replied:
i linked to the IAM documentation and indicated programmatic access.

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+
+1. Get a personal access token from [GitHub](https://github.com/settings/tokens) with repo and admin:repo_hook scopes.
+2. Navigate to the [AWS Console](https://console.aws.amazon.com) and create an IAM User with the following managed policies: AmazonS3FullAccess, AWSCodePipelineFullAccess, AWSCodeBuildAdminAccess.
+3. For this user, create security credentials and then register them in the `code-pipelines` profile on your computer with `aws configure --profile code-pipelines`.
```
listed aws cli as a requirement. added a link to the region table and suggested using the json output format.

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+3. For this user, create security credentials and then register them in the `code-pipelines` profile on your computer with `aws configure --profile code-pipelines`.
+4. Create a service role for CodeBuild as described [here](http://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html#setting-up-service-role).
+5. Create an IAM policy for a CodePipelines service role as described [here](http://docs.aws.amazon.com/codepipeline/latest/userguide/iam-identity-based-access-control.html#view-default-service-role-policy).
+6. Create a service role for CodePipelines by using the policy you created in step 5 and then set the STS trust to codepipeline.amazonaws.com, like what you did for CodeBuild.
```
While the AWS CLI can be used to automate these steps, it is preferable that the user manage all IAM configuration.

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+export AWS_ACCOUNT_NUMBER='000000000000'
+export AWS_CODEBUILD_ROLE='codebuild_role_name'
+export AWS_CODEPIPELINE_ROLE='codepipeline_role_name'
+export GITHUB_TOKEN='aaaa'
```
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,57 @@
+--bucket jgamcp \
+--github-owner amcp \
+--github-repo janusgraph \
+--github-branch codepipelines \
```
fixed

amcp commented on janusgraph-codepipelines-ci/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM maven:3-jdk-8
+ENV DEBIAN_FRONTEND noninteractive
+COPY settings-docker.xml /usr/share/maven/ref/
+ENV COMMIT 9165ee8407b3607f39031cdaeff1d15ba0fd4e9e
```
It is a recent commit I can bake into a Docker repository so that building code does not take a long time - I use it to pull all the maven dependencies. Any better ideas?

amcp commented on janusgraph-codepipelines-ci/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+         --batch-mode \
+         --file ${SRC_DIR}/pom.xml \
+         -DskipTests=true -Dmaven.javadoc.skip=true -Dmaven.compiler.showWarnings=false \
+    && rm -rf ${SRC_DIR}
```
The purpose of this Dockerfile is to create a reusable cache of maven dependencies.

amcp replied:
on second thought, it seems like I am not using this Dockerfile, so I will remove it.

amcp replied:
removed

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,83 @@
+[here](http://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html#setting-up-service-role).
+5. Create an IAM policy for a CodePipelines service role as described
+[here](http://docs.aws.amazon.com/codepipeline/latest/userguide/iam-identity-based-access-control.html#view-default-service-role-policy).
+6. Create a service role for CodePipelines by using the policy you created in step 5 and then set the STS trust to
```
Combine 5 and 6 (makes it consistent with step 4): "Create a CodePipeline [policy](http://docs.aws.amazon.com/codepipeline/latest/userguide/iam-identity-based-access-control.html#view-default-service-role-policy) and associate it to a new service role." Indicate (maybe as a sub-bullet) that "EC2" (or whichever) role be selected when creating the role since AWS CodePipeline doesn't appear to be an available option. Be more specific (maybe also as a sub-bullet) on how to set STS trust to codepipeline.amazonaws.com.

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,83 @@
+#define variables
+#GitHub personal access token you created in step 1
+export GITHUB_TOKEN=''
+#GitHub organization or user name
```
Add "of the repository to build". Need to make it clear that it's the user/group of the repository to test not (necessarily) the GitHub handle.

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,83 @@
+```
+
+3. Kick off the regular and TinkerPop tests. Use `export PIPELINE_CONFIGURATION=${PWD}/pipe.json` for regular tests
+and `export PIPELINE_CONFIGURATION=${PWD}/tp-pipe.json` for TinkerPop tests.
````
I think you can remove the `${PWD}/` prefixes

amcp replied:
removed

amcp commented on janusgraph-codepipelines-ci/settings-docker.xml:
```diff
@@ -0,0 +1,6 @@
+          xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
+                      https://maven.apache.org/xsd/settings-1.0.0.xsd">
+    <localRepository>/usr/share/maven/ref/repository</localRepository>
+</settings>
```
removed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,83 @@
+and a default [output format](http://docs.aws.amazon.com/cli/latest/userguide/controlling-output.html#controlling-output-format)
+of your choice (`json`, `text`, and `table` are available).
+4. Create a service role for CodeBuild as described
+[here](http://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html#setting-up-service-role).
```
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,83 @@
+mvn clean package
+#constants
+export AWS_PROFILE_NAME='code-pipelines'
+export AWS_ACCOUNT_NUMBER=`aws sts get-caller-identity --output text | cut -f1`
```
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,83 @@
+2. Second, configure the parameters below to customize your test stack.
+
+```bash
+#define variables
````
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,83 @@
+and `export PIPELINE_CONFIGURATION=${PWD}/tp-pipe.json` for TinkerPop tests.
+
+```bash
+#kick off regular tests
````
removed

amcp commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,83 @@
+--pipelines ${PIPELINE_CONFIGURATION}
+```
+
+4. Navigate to the [AWS Console](https://console.aws.amazon.com) and check on the status of your pipeline to see status.
````
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,83 @@
+#Name of bucket you want to use or create and use for storing your build artifacts.
+export AWS_S3_BUCKET_NAME=''
+#Name of pipeline configuration file to use
+export PIPELINE_CONFIGURATION=''
```
Created a "Create test pipeline" section and described one example

amcp commented:
I forgot to add a cleanup section. I can do that next.

amcp commented:
@sjudeng I added the cleanup section.

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,122 @@
+export AWS_REGION_NAME=''
+#Name of bucket you want to use or create and use for storing your build artifacts.
+export AWS_S3_BUCKET_NAME=''
+#Name of pipeline configuration file to use
```
"Name of pipeline **JSON** configuration file to use"

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,122 @@
+
+To kick off the regular and TinkerPop tests. Use `export PIPELINE_CONFIGURATION=pipe.json` for regular tests
+and `export PIPELINE_CONFIGURATION=tp-pipe.json` for TinkerPop tests.
+
```
Add a section break here for `## Run Test Pipelines`. I'd keep the "export PIPELINE_CONF..." examples in the previous section.

amcp replied:
fixed

sjudeng commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,124 @@
+```
+The first pipeline is called `j1` and the second pipeline is called `j2`. Each of these pipelines have four
+parallel build actions defined. Each build action is keyed at the action name and includes the environment
+variables to be passed CodeBuild and used in the buildspec.yml file. Use `export PIPELINE_CONFIGURATION=pipe.json` for regular tests
````
Instead of "Use 'export PIPELINE..." how about something like below?

Define the `PIPELINE_CONFIGURATION` environment variable to point to the custom pipeline file or use existing pipe.json/tp-pipe.json to run all default/TinkerPop tests.

```bash
export PIPELINE_CONFIGURATION=pipe.json
```

amcp replied:
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
````diff
@@ -0,0 +1,124 @@
+--pipelines ${PIPELINE_CONFIGURATION}
+```
+
+Navigate to the [AWS Console](https://console.aws.amazon.com/codepipeline) and check on the status of your pipeline to see status.
````
fixed

amcp commented on janusgraph-codepipelines-ci/README.md:
```diff
@@ -0,0 +1,124 @@
+4. Create an IAM policy for CodeBuild and associate it to a new service role as described
+[here](http://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html#setting-up-service-role). Select the
+__Amazon EC2__ role type as CodeBuild is not yet available on this page. Edit the new role's trust relationship
+and replace it with the text in step 16 of the CodeBuild user guide.
```
fixed

amcp commented:
lets create a separate issue for fixing ES build #372

amcp commented:
@sjudeng I [fixed assembly of composite](https://github.com/JanusGraph/janusgraph/pull/129/files#diff-c1b94d7778ab43a3cb43de588d9d5b39R14) builds like hadoop-lucene-solr so I am ready to close this PR out.
